server/shared/src/main/scala/org/http4s/server/middleware/EntityLimiter.scala

```diff
@@ -50,8 +50,10 @@ object EntityLimiter {
     _.pull
       .take(n)
       .flatMap {
-        case Some(_) => Pull.raiseError[F](EntityTooLarge(n))
-        case None => Pull.done
+        case Some(rest) => // if rest is empty then we won't raise an error
+          (rest >> Stream.raiseError(EntityTooLarge(n))).pull.echo
```

Comment:
Not sure if it's a better option, but if we would be more explicit here
```suggestion
        case Some(rest) =>
          rest.pull.take(1).flatMap {
            case Some(_) =>
              Pull.raiseError[F](EntityTooLarge(n))
            case None =>
              Pull.done
          }
```
we could give up on that comment :)

Reply:
Does that work? If you have `Stream.emit(0)` and you `take(1)`, you will succeed, and then the tail of the `Stream` will be `None`.

The `.pull.take(n)` API is confusing because the `Option[Stream]` it gives you is not the elements it took, but what is left of the stream _after_ emitting `n` elements.

Reply:
Checked now, and your tests work well with this patch 🤔

Reply:
@danicheg thanks, good catch. The difference is actually really subtle: I just added a test in 0bb0eeab4a467dca5f827396e0199fd24525cc5b that will fail for that patch. **tl;dr** we not only need to verify that the `EntityLimiter` raises an error if the number of bytes is exceeded, we also need to verify that it _stops_ passing those bytes to the `EntityDecoder` (or other consumer).

Reply:
> we also need to verify that it _stops_ passing those bytes to the `EntityDecoder` (or other consumer).

That's why this middleware raises and doesn't simply force returning a 413?

Reply:
Good question. I'm not sure if a middleware can actually do that.